src/psyclone/transformations.py

```diff
@@ -2443,20 +2446,55 @@ def validate(self, node, options=None):
                 raise TransformationError(
                     f"Failed to create PSyIR for kernel '{node.name}'. "
                     f"Cannot transform such a kernel.") from error
-
-            # Check that the kernel does not access any data or routines via a
-            # module 'use' statement
-            imported_variables = kernel_schedule.symbol_table.imported_symbols
-            if imported_variables:
+            k_or_r = "Kernel"
+        else:
+            # Supplied node is a PSyIR Routine which *is* a Schedule.
+            kernel_schedule = node
+            k_or_r = "routine"
+
+        # Check that the routine does not access any data that is imported via
+        # a 'use' statement.
+        refs = kernel_schedule.walk(Reference)
+        for ref in refs:
+            if ref.symbol.is_import:
+                # resolve_deferred does nothing if the Symbol type is known.
+                try:
+                    ref.symbol.resolve_deferred()
+                except SymbolError:
+                    # TODO #11 - log that we failed to resolve this Symbol.
+                    pass
+                if (isinstance(ref.symbol, DataSymbol) and
+                        ref.symbol.is_constant):
+                    # An import of a compile-time constant is fine.
+                    continue
+                raise TransformationError(
+                    f"{k_or_r} '{node.name}' accesses the symbol "
+                    f"'{ref.symbol}' which is imported. If this symbol "
+                    f"represents data then it must first be converted to a "
+                    f"{k_or_r} argument using the KernelImportsToArguments "
+                    f"transformation.")
+
+        # Check any accesses within CodeBlocks.
+        cblocks = kernel_schedule.walk(CodeBlock)
+        for cblock in cblocks:
+            names = cblock.get_symbol_names()
+            for name in names:
+                sym = kernel_schedule.symbol_table.lookup(name)
+                if sym.is_import:
+                    raise TransformationError(
+                        f"{k_or_r} '{node.name}' accesses the symbol "
+                        f"'{sym.name}' within a CodeBlock and this symbol is "
+                        f"imported. 'ACC routine' cannot be added to such a "
+                        f"{k_or_r}.")
```

Comment:
Maybe if there is any Codeblock at all, regardless of the symbols used, it should not validate as we can not guarantee that is going to work on the GPU.

Which mean that maybe we just need to add CodeBlock in the class exclude_node_type instead of this check?

Reply:
Unfortunately this transformation is not a `RegionTrans` so there's no exclude list. Therefore I've just altered the validate to reject CodeBlocks (unless the user forces it).